saml2/saml2-service-provider/src/main/java/org/springframework/security/saml2/serviceprovider/registration/RelyingPartyRegistration.java

```diff
@@ -0,0 +1,205 @@
+		return this.remoteIdpEntityId;
+	}
+
+	public String getRegistrationId() {
+		return this.registrationId;
+	}
+
+	public String getAssertionConsumerServiceUrlTemplate() {
+		return this.assertionConsumerServiceUrlTemplate;
+	}
+
+	public String getIdpWebSsoUrl() {
+		return this.idpWebSsoUrl;
+	}
+
+	public String getLocalEntityIdTemplate() {
+		return this.localEntityIdTemplate;
+	}
+
+	public List<Saml2X509Credential> getCredentialsForUsage(Saml2X509CredentialUsage... types) {
+		if (types == null || types.length == 0) {
+			return this.credentials;
+		}
+		Set<Saml2X509CredentialUsage> typeset = new HashSet<>(asList(types));
+		List<Saml2X509Credential> result = new LinkedList<>();
+		for (Saml2X509Credential c : this.credentials) {
+			if (containsCredentialForTypes(c.getSaml2X509CredentialUsages(), typeset)) {
+				result.add(c);
+			}
+		}
+		return result;
+	}
+
+	private boolean containsCredentialForTypes(Set<Saml2X509CredentialUsage> existing, Set<Saml2X509CredentialUsage> requested) {
+		for (Saml2X509CredentialUsage u : requested) {
+			if (existing.contains(u)) {
+				return true;
+			}
+		}
+		return false;
+	}
+
+	public static Builder withRegistrationId(String registrationId) {
+		Assert.hasText(registrationId, "registrationId cannot be empty");
+		return new Builder(registrationId);
+	}
+
+	public static class Builder {
+		private String registrationId;
+		private String remoteIdpEntityId;
+		private String idpWebSsoUrl;
+		private String assertionConsumerServiceUrlTemplate;
+		private List<Saml2X509Credential> credentials = new LinkedList<>();
+		private String localEntityIdTemplate = "{baseUrl}/saml2/service-provider-metadata/{registrationId}";
+
+		private Builder(String registrationId) {
+			this.registrationId = registrationId;
+		}
+
+		public Builder registrationId(String id) {
+			this.registrationId = id;
+			return this;
+		}
+
+		public Builder remoteIdpEntityId(String entityId) {
+			this.remoteIdpEntityId = entityId;
+			return this;
+		}
+
+		/**
+		 * <a href="https://wiki.shibboleth.net/confluence/display/CONCEPT/AssertionConsumerService">Assertion Consumer
+		 * Service</a> URL template. It can contain variables {@code baseUrl}, {@code registrationId},
+		 * {@code baseScheme}, {@code baseHost}, and {@code basePort}.
+		 * @param assertionConsumerServiceUrlTemplate the Assertion Consumer Service URL template (i.e.
+		 * "{baseUrl}/login/saml2/sso/{registrationId}".
+		 * @return the Builder
+		 */
+		public Builder assertionConsumerServiceUrlTemplate(String assertionConsumerServiceUrlTemplate) {
+			this.assertionConsumerServiceUrlTemplate = assertionConsumerServiceUrlTemplate;
+			return this;
+		}
+
+		public Builder idpWebSsoUrl(String url) {
+			this.idpWebSsoUrl = url;
+			return this;
+		}
+
+		public Builder credential(Saml2X509Credential credential) {
+			this.credentials.add(credential);
+			return this;
+		}
+
+		public Builder credentials(Collection<Saml2X509Credential> credentials) {
```

Comment:
It would be more powerful to have this take a `Consumer<Collection<Saml2X509Credential>>` in the same way that `Jwt.Builder#claims` or `WebClient.Builder#headers` does.

Reply:
Consumer added
569082c